core/src/test/java/org/apache/accumulo/core/data/MutationTest.java

```diff
@@ -187,7 +187,8 @@ private Value nv(String s) {
   public void testPuts() {
     Mutation m = new Mutation(new Text("r1"));
 
-    m.put(nt("cf1"), nt("cq1"), nv("v1"));
+    m.at().family(nt("cf1")).qualifier(nt("cq1")).set(nv("v1"));
```

Comment:
More test are needed.  Need to test all of the types and test adding multiple columns.  I would recommend doing this in separate test.  Could do something like the following.

```java
 @Test
  public void testAtTypes() {
    String fam = "f145";
    String qual = "q452";
    String val = "v52093209284";

    Mutation expected = new Mutation("row5");
    expected.put(fam, qual, val);

    //Test all family methods
    Mutation actual = new Mutation("row5");
    actual.at().family(fam).qualifier(qual).set(val);
    assertEquals(expected, actual);

    actual = new Mutation("row5");
    actual.at().family(fam.getBytes(UTF_8)).qualifier(qual).set(val);
    assertEquals(expected, actual);

    actual = new Mutation("row5");
    actual.at().family(new Text(fam)).qualifier(qual).set(val);
    assertEquals(expected, actual);

    ByteBuffer bbFam = ByteBuffer.wrap(fam.getBytes(UTF_8));
    actual = new Mutation("row5");
    actual.at().family(bbFam).qualifier(qual).set(val);
    assertEquals(expected, actual);
    //TODO ensure that bbFam position was not changed

    //Test all qualifier methods
    actual = new Mutation("row5");
    actual.at().family(fam).qualifier(qual.getBytes(UTF_8)).set(val);
    assertEquals(expected, actual);

    actual = new Mutation("row5");
    actual.at().family(fam).qualifier(new Text(qual)).set(val);
    assertEquals(expected, actual);

    ByteBuffer bbQual = ByteBuffer.wrap(qual.getBytes(UTF_8));
    actual = new Mutation("row5");
    actual.at().family(fam).qualifier(bbQual).set(val);
    assertEquals(expected, actual);
    //TODO ensure bbQual position was not changed

    //Test all visibility methods
    //TODO

    //Test all set method
    //TODO

    // The following will not compile because set returns void.  It would make these test a lot shorter if set() returned the
    // mutation.  However, not sure this is a good idea.  Would not work with subclasses like ConditionalMutation.
    assertEquals(expected, new Mutation("row5").at().family(fam).qualifier(qual).set(val));
  }

  @Test
  public void testAtMultiple() {
    ColumnVisibility cv = new ColumnVisibility("A&B");

    Mutation expected = new Mutation("row5");
    expected.put("fam123", "q77", cv, "v4224");
    expected.put("fam123", "q47", 56, "v345");
    expected.put("fam999", "q77", "v1");
    expected.putDelete("fam89", "q69");

    Mutation actual = new Mutation("row5");
    actual.at().family("fam123").qualifier("q77").visibility(cv).set("v4224");
    actual.at().family("fam123").qualifier("q47").timestamp(56).set("v345");
    actual.at().family("fam999").qualifier("q77").set("v1");
    actual.at().family("fam89").qualifier("q69").delete();

    assertEquals(expected, actual);
  }
```

Writing these test made me wonder if the `set()` methods should have a return type of `Mutation`.  I am not sure about this, need to think through the implications for ConditionalMutation.

Reply:
For sure needs more testing. I'm going to try out the design of having the `set()` methods return `Mutation` in this next commit, in addition to more tests. Thank you for the great start you've given me here.

Reply:
Idea: add a new group of `set` methods called `conditionalSet` that return a `ConditionalMutation` object. This will enable users to continue the chain and add conditions as well.

`public ConditionalMutation conditionalSet(byte[] val) { }`

`ConditionalMutation actual = new ConditionalMutation("row5");
actual.at().family("fam123").qualifier("q77").visibility(cv)
.conditionalSet("v4224").addCondition(new Condition())`

Reply:
Where were you thinking of adding the `conditionalSet()` method?